### PR TITLE
Optimizing ML Models with Optuna

### LINKED ISSUE #5749: Optimizing ML Models with Optuna
### Topic Suggestion
Optimizing ML Models with Optuna
#### Proposed article introduction
Optuna is the SOTA algorithm used to make changes to ML and deep learning models. It all depends on the Bayesian fine-tuning method. It stops trying combinations that don't improve our score any more and only tries combinations that do.

People who work on ML models have a hard time. In order to make it easier for us to fine-tune our model, there are a lot of tools out there. You need to make sure that our data is important and that our ML model fits on it perfectly. Here, we'll learn how to make an Xgboost regressor work even better.

When you use machine learning, a hyperparameter is a parameter that can be changed to change how the learning process works.  The values of hyperparameters can't be guessed from the data. It's important to set these parameters so that the model can learn from our data. You can change the learning parameters so that the model fits our data and learns from it to make smart predictions.
#### Key takeaways 
- Visualization of Tuning
- Comparing the Tuned Model and Base Model
- Creating Hyper-Parameters Space

#### Article quality
In this article, we'll learn how to improve ML models and how to find the best parameters for the models we make. The code will be easy to read and understand.
#### References
N/A
### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by NdoloBonface**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5749
> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by srishilesh** [comment]
Hey @NdoloBonface ,

Thanks for reaching out to me.

I've some PRs to review in my backlog. Please wait until a peer-reviewer pings you with the changes to be made.

Thanks.

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
+It is possible to use various resources to improve our model. Ensure that the ML model is acceptable for our data and required. The following section will discuss enhancements and new applications for `Xgboost` regressors.
+
+If you use machine learning methods, a `hyperparameter` is a parameter that may be changed to influence the learning process. Using our data, we'll tweak the model's learning parameters.
+### Optuna functionalities
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:4]: Add `Table of contents` and `Pre-requisites` sections. Refer my articles.

Comment (srishilesh) [reply content/articles/optimizing-ml-models-with-optuna/index.md:4]: State what optuna is, why we use it, where we use it, how it helps, what are the current state methods. Add reference links.

Have this as a separate topic. Cover this in-detail

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
+It is possible to use various resources to improve our model. Ensure that the ML model is acceptable for our data and required. The following section will discuss enhancements and new applications for `Xgboost` regressors.
+
+If you use machine learning methods, a `hyperparameter` is a parameter that may be changed to influence the learning process. Using our data, we'll tweak the model's learning parameters.
+### Optuna functionalities
+- Installation is quick and easy, and there are only a few prerequisites to meet.
+- Python's conditionals and loops may be used to define search areas.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:6]: Explain more about this.

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+dataframe['target'] = Our_loadboston.target
+X = dataframe.iloc[:,dataframe.columns != 'target']
+y = dataframe.target
+# A pandas data frame will be used to segregate features and targets when the data is returned as a matrices by load boston().
+from sklearn.preprocessing import StandardScaler
+standardSc = StandardScaler()
+X = standardSc.fit_transform(X)
+# By dividing the dataset into train and test, the model is better able to reach a convergent state.
+from sklearn.model_selection import train_test_split
+X_train, X_test, y_train,y_test = train_test_split(X, y, test_size = 0.2, random_state = 12)
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:30]: You can break down this into multiple code snippets like:
- Loading dataset
- Standard scaling the dataset
- Dataset split

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+```
+#### Testing the model
+It's easier to compare results when you've already built a baseline model. We can see how much better our model is after fine-tuning the parameters.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:33]: How easy it is? Stating them just in words doesn't justify your point. State proper reference links.

If possible, build some other model and then compare them here.

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import xgboost as gbst
+from sklearn.model_selection import cross_val_score
+xgboo_reg = gbst.XGBRegressor()
+# Cross-validation is a method for evaluating a score
+ourScores = cross_val_score(xgboo_reg, X_train,y_train , scoring = 'neg_root_mean_squared_error',
+                            #using all available CPUs
+                            n_jobs = -1,cv = 10)
+print(num.mean(ourScores), num.std(ourScores))
+print(ourScores)
+```
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:44]: What are we doing here? Why are learning this? How is it related to optuna?

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+X = dataframe.iloc[:,dataframe.columns != 'target']
+y = dataframe.target
+# A pandas data frame will be used to segregate features and targets when the data is returned as a matrices by load boston().
+from sklearn.preprocessing import StandardScaler
+standardSc = StandardScaler()
+X = standardSc.fit_transform(X)
+# By dividing the dataset into train and test, the model is better able to reach a convergent state.
+from sklearn.model_selection import train_test_split
+X_train, X_test, y_train,y_test = train_test_split(X, y, test_size = 0.2, random_state = 12)
+```
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:31]: What are we doing here? Why are learning this? How is it related to optuna?

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
+It is possible to use various resources to improve our model. Ensure that the ML model is acceptable for our data and required. The following section will discuss enhancements and new applications for `Xgboost` regressors.
+
+If you use machine learning methods, a `hyperparameter` is a parameter that may be changed to influence the learning process. Using our data, we'll tweak the model's learning parameters.
```
[the remark this answers is not in the document]
Comment (srishilesh) [reply content/articles/optimizing-ml-models-with-optuna/index.md:3]: Explain about XGB and hyperparameter tuning in the later sections. Give a basic overview of them

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+print(num.mean(ourScores), num.std(ourScores))
+print(ourScores)
+```
+Output:
+```bash
+-3.0784942308511307 0.42284534035667176
+[-2.62847993 -3.60582341 -3.31998107 -3.60355075 -3.03760446 -2.50243868 -2.92302011 -2.61990331 -3.67855489 -2.8655857 ]
+```
+#### Defining Functions
+`return_score()` provides a cross-verified score based on keyword arguments sent to it. We will take 1000 samples.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:51]: Why are we doing this? Explain in-detail

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+```python
+def return_score(param):
+  ouNewModel = gbst.XGBRegressor(**param)  
+  regressionModelse = -num.mean(cross_val_score(ouNewModel,X_train[:1000],y_train[:10000], cv = 4, n_jobs =-1,scoring='neg_root_mean_squared_error'))
+  return regressionModelse
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:56]: Explain the code snippets as bulleted points.

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+```
+#### Importing optuna
+This is where we import Optuna.
+
+An objective function may be examined via a trial to see whether it successfully tunes.
+```python
+import optuna
+# The Samplers class defines our hyper-parameter space.
+from Optuna Import Trial, visualization
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:65]: `Import`? or `import`?

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+#### Defining Functions
+`return_score()` provides a cross-verified score based on keyword arguments sent to it. We will take 1000 samples.
+```python
+def return_score(param):
+  ouNewModel = gbst.XGBRegressor(**param)  
+  regressionModelse = -num.mean(cross_val_score(ouNewModel,X_train[:1000],y_train[:10000], cv = 4, n_jobs =-1,scoring='neg_root_mean_squared_error'))
+  return regressionModelse
+```
+#### Importing optuna
+This is where we import Optuna.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:59]: Talk about pip installation too

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+An objective function may be examined via a trial to see whether it successfully tunes.
+```python
+import optuna
+# The Samplers class defines our hyper-parameter space.
+from Optuna Import Trial, visualization
+from optuna.samplers import TPESampler
+```
+#### Creating the primary objective function
+We place all the possible hyper-parameter ranges, and the method accepts a trial object and should return a score value.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:69]: Explain in-detail. The objective is not clear

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+                'min_child_weight':trial.suggest_int('min_child_weight',0,5),
+                'learning_rate':trial.suggest_loguniform('learning_rate',0.05,0.5),
+               #log distribution between [0.05,.5]
+                'colsample_bytree':trial.suggest_uniform('colsample_bytree',0.4,0.9),
+                'max_depth':trial.suggest_int('max_depth', 3, 5),
+                'nthread' : -1
+            }
+  #returns the regressionModelse score
+  return(return_score(parameter))
+```
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:88]: What are we doing here? Explain them in-detail. Why are doing this? How is it related?

Explain them as bulleted points.

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+#### Creating an optuna–study object
+This study object stores all the information about the hyper-parameters tested under that research.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:90]: What is it? Explain

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+```python
+studyObject1 = optuna.create_study(
+    direction='minimize',sampler=TPESampler())
+    # Bayesian Sampling Technique
+studyObject1.optimize(objective, n_trials= 500)
+```
+After running i found that my best trial was 459.
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:97]: What are we doing here?

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+Output:
+
+![Best trial](/engineering-education/optimizing-ml-models-with-optuna/trial.png)
+#### Visualizing search history
+Our model can readily be shown to be affected by hyper-parameters by looking at the search history. It is possible to further decrease the scope of our search by cutting down the search parameters.
+```python
+optuna.visualization.plot_slice(studyObject1)
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:105]: Decrease search? Explanation is not enough

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+ 'n_estimators': 475,
+ 'ourlearning_rate': 0.053232877076795374,
+ 'reg_alpha': 1.908137784762065,
+ 'reg_lambda': 0.03170430168340592,
+ 'subsample': 0.7131725068799382}
+ ```
+Let's examine how far we've come with our model:
+```python
+parameter = {}
+print(f"without optimization {return_score(parameter)}")
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:129]: Need more explanation. I can't really get the context of what is being explained.

File: content/articles/optimizing-ml-models-with-optuna/index.md [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+> By restricting the ranges of hyperparameters, we can fine-tune even more.
+### Reasons for using Optuna
+- The length of the optimization procedure may be specified.
+- Pandas DataFrame integration.
+- Pruning is a technique that the algorithm uses to eliminate poor-quality samples quickly.
+
+The Optuna software framework automates the hyperparameter optimization process. **Tree-structured Parzen Estimator** is the name of the `Bayesian` optimization algorithm it uses. According to Bayesian fine-tuning, once it stops improving our score anymore, it stops looking for new ones.
+
+Optuna can be used with any Python framework, including Chainer, Scikit-learn, Pytorch, and others, because it is framework agnostic. In PFN projects, Optuna is used successfully.
+### Conclusion
```
Comment (srishilesh) [inline content/articles/optimizing-ml-models-with-optuna/index.md:146]: Needs in-depth explanation

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @NdoloBonface ,

On going through your article, I found that your article lacked clarity in explaining concepts. As a beginner in learning optuna, I was unable to follow up the concepts. Moreover, you haven't stated the objectives of learning it, why we use it, what are the existing methods, how is it better than them, and so on.

Even the code snippets lacked clarity and objective. You haven't explained them.

I've pointed out a few suggestions. Please go through them carefully and address all of them. I would highly suggest you to go through approved/published articles to see how other authors have formatted and structured their articles.

Since your PR requires too many changes, we will closing it for now. You may create a new PR and ping me there, only when you are done addressing all the changes.

Let me know if you have any questions.

Thanks.

**Comment by NdoloBonface** [comment]
@srishilesh I will work on it and ping you right away. Thanks.
I was also working on the PR the moment you closed it.

File: content/articles/optimizing-ml-models-with-optuna/index.md (hunk 1 above)
Comment (NdoloBonface) [reply content/articles/optimizing-ml-models-with-optuna/index.md:4]: Already covered in the later section

**Pull request closed without merging**